contrib/mesos/pkg/queue/delay/indexed_delay_queue.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package delay
+
+import (
+	"container/heap"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pivotal-golang/clock"
+
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue/priority"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// IndexedDelayQueue is a thread-safe, time-based queue with uniquely identified items.
+//
+// Adding or offering an item with an ID matching an item still in the queue will
+// invoke the configured SchedulingPolicy and supplied ValueReplacementPolicy to
+// determine the desired behavior.
+//
+// Items can only be popped after their event time has been reached.
+// Use Add to specify a delay duration and have the event time calculated.
+// Use Offer to specify a specific event time.
+type IndexedDelayQueue struct {
+	*Queue
```

Comment:
you're right, there's not. the original implementation used a named field
`delegate`

On Mon, Sep 28, 2015 at 7:22 PM, Karl Isenberg <notifications@github.com>
wrote:

> In contrib/mesos/pkg/queue/delay/indexed_delay_queue.go
> <https://github.com/kubernetes/kubernetes/pull/14610#discussion_r40618264>
> :
>
> > +	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
> > +	"k8s.io/kubernetes/contrib/mesos/pkg/queue/priority"
> > +	"k8s.io/kubernetes/pkg/util/sets"
> > +)
> > +
> > +// IndexedDelayQueue is a thread-safe, time-based queue with uniquely identified items.
> > +//
> > +// Adding or offering an item with an ID matching an item still in the queue will
> > +// invoke the configured SchedulingPolicy and supplied ValueReplacementPolicy to
> > +// determine the desired behavior.
> > +//
> > +// Items can only be popped after their event time has been reached.
> > +// Use Add to specify a delay duration and have the event time calculated.
> > +// Use Offer to specify a specific event time.
> > +type IndexedDelayQueue struct {
> > +	*Queue
>
> I don't think there's a way to not expose the Queue when it's an anonymous
> field. If I make it a named field then IndexedDelayQueue won't inherent the
> functions and fields of Queue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14610/files#r40618264>.
>